In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
import pandas as pd
import tensorflow as tf

# read the charity_data.csv.
application_df = pd.read_csv("./Resources/model_charity_data.csv")
application_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                     1.0   
1       1   108590              1                     0.0   
2       1     5000              0                     0.0   
3       1     6692              1                     0.0   
4       1   142590              1                     0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                   0.0                  0.0                  0.0   
1                   0.0                  1.0                  0.0   
2                   0.0                  0.0                  0.0   
3                   0.0                  1.0                  0.0   
4                   0.0                  1.0                  0.0   

   APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  AFFILIATION_CompanySponsored  \
0                  0.0                  0.0                           0.0   
1                  0.0                  0.0                           0.0   
2                  1.0                  0.0                           1.0   
3                  0.0                  0.0                           1.0   
4                  0.0                  0.0                           0.0   

   ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0  ...                0.0                     0.0                       0.0   
1  ...                1.0                     0.0                       0.0   
2  ...                0.0                     0.0                       0.0   
3  ...                0.0                     1.0                       0.0   
4  ...                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 41 columns]

In [2]:
application_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34299 entries, 0 to 34298
Data columns (total 41 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   STATUS                        34299 non-null  int64  
 1   ASK_AMT                       34299 non-null  int64  
 2   IS_SUCCESSFUL                 34299 non-null  int64  
 3   APPLICATION_TYPE_Other        34299 non-null  float64
 4   APPLICATION_TYPE_T19          34299 non-null  float64
 5   APPLICATION_TYPE_T3           34299 non-null  float64
 6   APPLICATION_TYPE_T4           34299 non-null  float64
 7   APPLICATION_TYPE_T5           34299 non-null  float64
 8   APPLICATION_TYPE_T6           34299 non-null  float64
 9   AFFILIATION_CompanySponsored  34299 non-null  float64
 10  AFFILIATION_Family/Parent     34299 non-null  float64
 11  AFFILIATION_Independent       34299 non-null  float64
 12  AFFILIATION_National          34299 non-null  float64
 13  A

In [3]:
application_df.nunique()

STATUS                             2
ASK_AMT                         8747
IS_SUCCESSFUL                      2
APPLICATION_TYPE_Other             2
APPLICATION_TYPE_T19               2
APPLICATION_TYPE_T3                2
APPLICATION_TYPE_T4                2
APPLICATION_TYPE_T5                2
APPLICATION_TYPE_T6                2
AFFILIATION_CompanySponsored       2
AFFILIATION_Family/Parent          2
AFFILIATION_Independent            2
AFFILIATION_National               2
AFFILIATION_Other                  2
AFFILIATION_Regional               2
CLASSIFICATION_C1000               2
CLASSIFICATION_C1200               2
CLASSIFICATION_C2000               2
CLASSIFICATION_C2100               2
CLASSIFICATION_C3000               2
CLASSIFICATION_Other               2
USE_CASE_CommunityServ             2
USE_CASE_Heathcare                 2
USE_CASE_Other                     2
USE_CASE_Preservation              2
USE_CASE_ProductDev                2
ORGANIZATION_Association           2
O

In [4]:
# drop the "ASK_AMT" column which is the feature with the most noise.
application_df = application_df.drop(columns = "ASK_AMT")
application_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34299 entries, 0 to 34298
Data columns (total 40 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   STATUS                        34299 non-null  int64  
 1   IS_SUCCESSFUL                 34299 non-null  int64  
 2   APPLICATION_TYPE_Other        34299 non-null  float64
 3   APPLICATION_TYPE_T19          34299 non-null  float64
 4   APPLICATION_TYPE_T3           34299 non-null  float64
 5   APPLICATION_TYPE_T4           34299 non-null  float64
 6   APPLICATION_TYPE_T5           34299 non-null  float64
 7   APPLICATION_TYPE_T6           34299 non-null  float64
 8   AFFILIATION_CompanySponsored  34299 non-null  float64
 9   AFFILIATION_Family/Parent     34299 non-null  float64
 10  AFFILIATION_Independent       34299 non-null  float64
 11  AFFILIATION_National          34299 non-null  float64
 12  AFFILIATION_Other             34299 non-null  float64
 13  A

In [5]:
# Split our preprocessed data into our features and target arrays
X = application_df.drop(columns="IS_SUCCESSFUL")
y = application_df["IS_SUCCESSFUL"]

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=78)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(25724, 39) (8575, 39) (25724,) (8575,)


In [6]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [7]:
# Define the model - add a second hidden layer with ten nodes to see if it improves model accuracy

number_input_features = len(X_train_scaled[0])
hidden_layer1 = 20
hidden_layer2 = 10

nn_optimization1 = tf.keras.models.Sequential()

# First hidden layer
nn_optimization1.add(tf.keras.layers.Dense(input_dim=number_input_features, activation="relu", units=hidden_layer1))

# Second hidden layer
nn_optimization1.add(tf.keras.layers.Dense(activation="relu", units=hidden_layer2))

# Output layer
nn_optimization1.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_optimization1.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 20)                800       
                                                                 
 dense_1 (Dense)             (None, 10)                210       
                                                                 
 dense_2 (Dense)             (None, 1)                 11        
                                                                 
Total params: 1,021
Trainable params: 1,021
Non-trainable params: 0
_________________________________________________________________


In [8]:
# Import checkpoint dependencies
import os
from tensorflow.keras.callbacks import ModelCheckpoint

# Define the checkpoint path and filenames
os.makedirs("checkpoints_optimized1/",exist_ok=True)
checkpoint_path = "checkpoints_optimized1/weights.{epoch:02d}.hdf5" # https://keras.io/api/callbacks/model_checkpoint/

In [9]:
# Compile the model
nn_optimization1.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])
           
# Create a callback object that saves the model's weights every epoch
cp_callback = ModelCheckpoint(
        filepath = checkpoint_path,
        verbose=1,
        save_weights_only=True,
        save_freq=5
)

In [10]:
# Train the model
nn_optimization1.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 4s 3ms/step - loss: 0.5869 - accuracy: 0.7158
Epoch 2/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5628 - accuracy: 0.7300
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5585 - accuracy: 0.7311
Epoch 4/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5568 - accuracy: 0.7318
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5545 - accuracy: 0.7330
Epoch 6/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5537 - accuracy: 0.7330
Epoch 7/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5531 - accuracy: 0.7332
Epoch 8/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5525 - accuracy: 0.7334
Epoch 9/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5517 - accuracy: 0.7334
Epoch 10/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5512 - accura

In [11]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_optimization1.evaluate(X_test_scaled,y_test,verbose=2, callbacks= [cp_callback])
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5555 - accuracy: 0.7279 - 1s/epoch - 4ms/step
Loss: 0.5555292367935181, Accuracy: 0.7279300093650818


* By adding a second hidden layer with five nodes, the accuracy marginally increased from 72.55% to 72.79%.

In [12]:
# Define a new deep neural network model adding more nodes to the first layer to see if it further improves accuracy
number_input_features = len(X_train_scaled[0])
hidden_layer1 = 25
hidden_layer2 = 10

nn_optimization2 = tf.keras.models.Sequential()

# First hidden layer
nn_optimization2.add(tf.keras.layers.Dense(input_dim=number_input_features, activation="relu", units=hidden_layer1))

# Second hidden layer
nn_optimization2.add(tf.keras.layers.Dense(activation="relu", units=hidden_layer2))

# Output layer
nn_optimization2.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_optimization2.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 25)                1000      
                                                                 
 dense_4 (Dense)             (None, 10)                260       
                                                                 
 dense_5 (Dense)             (None, 1)                 11        
                                                                 
Total params: 1,271
Trainable params: 1,271
Non-trainable params: 0
_________________________________________________________________


In [13]:
# Import checkpoint dependencies
import os
from tensorflow.keras.callbacks import ModelCheckpoint

# Define the checkpoint path and filenames
os.makedirs("checkpoints_optimized2/",exist_ok=True)
checkpoint_path = "checkpoints_optimized2/weights.{epoch:02d}.hdf5" # https://keras.io/api/callbacks/model_checkpoint/

In [14]:
# Compile the model
nn_optimization2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [15]:
# Train the model
nn_optimization2.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 4s 3ms/step - loss: 0.5948 - accuracy: 0.7115
Epoch 2/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5605 - accuracy: 0.7306
Epoch 3/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5576 - accuracy: 0.7300
Epoch 4/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5558 - accuracy: 0.7295
Epoch 5/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5544 - accuracy: 0.7309
Epoch 6/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5534 - accuracy: 0.7317
Epoch 7/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5529 - accuracy: 0.7311
Epoch 8/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5520 - accuracy: 0.7319
Epoch 9/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5517 - accuracy: 0.7321
Epoch 10/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5514 - accura

In [16]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_optimization2.evaluate(X_test_scaled,y_test,verbose=2, callbacks= [cp_callback])
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5550 - accuracy: 0.7275 - 931ms/epoch - 3ms/step
Loss: 0.5549562573432922, Accuracy: 0.7274635434150696


* By increasing the number of nodes in the second hidden layer from five to ten, the accuracy marginally increased from the original model from 72.55% to 72.75%.

In [17]:
# Define a new deep neural network model, adding another layer with 8 nodes
number_input_features = len(X_train_scaled[0])
hidden_layer1 = 25
hidden_layer2 = 10
hidden_layer3 = 8

nn_optimization3 = tf.keras.models.Sequential()

# First hidden layer
nn_optimization3.add(tf.keras.layers.Dense(input_dim=number_input_features, activation="relu", units=hidden_layer1))

# Second hidden layer
nn_optimization3.add(tf.keras.layers.Dense(activation="relu", units=hidden_layer2))

# Third hidden layer
nn_optimization3.add(tf.keras.layers.Dense(activation="relu", units=hidden_layer3))

# Output layer
nn_optimization3.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn_optimization3.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 25)                1000      
                                                                 
 dense_7 (Dense)             (None, 10)                260       
                                                                 
 dense_8 (Dense)             (None, 8)                 88        
                                                                 
 dense_9 (Dense)             (None, 1)                 9         
                                                                 
Total params: 1,357
Trainable params: 1,357
Non-trainable params: 0
_________________________________________________________________


In [18]:
# Import checkpoint dependencies
import os
from tensorflow.keras.callbacks import ModelCheckpoint

# Define the checkpoint path and filenames
os.makedirs("checkpoints_optimized3/",exist_ok=True)
checkpoint_path = "checkpoints_optimized3/weights.{epoch:02d}.hdf5" # https://keras.io/api/callbacks/model_checkpoint/

In [19]:
# Compile the model
nn_optimization3.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [20]:
# Train the model
nn_optimization3.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 4s 4ms/step - loss: 0.5850 - accuracy: 0.7124
Epoch 2/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5598 - accuracy: 0.7299
Epoch 3/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5559 - accuracy: 0.7308
Epoch 4/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5545 - accuracy: 0.7315
Epoch 5/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5534 - accuracy: 0.7319
Epoch 6/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5526 - accuracy: 0.7315
Epoch 7/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5517 - accuracy: 0.7329
Epoch 8/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5509 - accuracy: 0.7336
Epoch 9/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5505 - accuracy: 0.7336
Epoch 10/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5503 - accura

In [21]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_optimization3.evaluate(X_test_scaled,y_test,verbose=2, callbacks= [cp_callback])
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5572 - accuracy: 0.7273 - 985ms/epoch - 4ms/step
Loss: 0.5571914911270142, Accuracy: 0.7273469567298889


* By increasing the number of nodes in the second hidden layer from five to ten, the accuracy marginally increased from the original model from 72.55% to 72.73%.

In [23]:
# Export most accurate optimized model weights to HDF5 file
nn_optimization1.save("AlphabetSoupCharity_Optimization.h5")